File: phases/07-transformers-deep-dive/02-self-attention-from-scratch/code/main.jl
# Self-attention from scratch in Julia. Scaled dot-product attention,
# numerically-stable row-wise softmax, single-head and multi-head
# self-attention. Stdlib only. Sources:
#   https://arxiv.org/abs/1706.03762
#   https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/
#   https://docs.julialang.org/en/v1/stdlib/Random/

using Random
using LinearAlgebra
using Printf


function softmax_rows(M::Matrix{Float64})::Matrix{Float64}
    out = similar(M)
    for i in 1:size(M, 1)
        row = M[i, :]
        m = maximum(row)
        e = exp.(row .- m)
        s = sum(e)
        out[i, :] = e ./ s
    end
    return out
end


function scaled_dot_product_attention(Q::Matrix{Float64}, K::Matrix{Float64},
                                      V::Matrix{Float64})
    dk = size(Q, 2)
    scores = (Q * transpose(K)) ./ sqrt(dk)
    weights = softmax_rows(scores)
    output = weights * V
    return output, weights
end


struct SelfAttention
    Wq::Matrix{Float64}
    Wk::Matrix{Float64}
    Wv::Matrix{Float64}
    dk::Int
end


function SelfAttention(d_model::Int, dk::Int, dv::Int; seed::Int=42)
    rng = MersenneTwister(seed)
    scale_qk = sqrt(2.0 / (d_model + dk))
    scale_v = sqrt(2.0 / (d_model + dv))
    Wq = scale_qk .* randn(rng, d_model, dk)
    Wk = scale_qk .* randn(rng, d_model, dk)
    Wv = scale_v .* randn(rng, d_model, dv)
    return SelfAttention(Wq, Wk, Wv, dk)
end


function forward(attn::SelfAttention, X::Matrix{Float64})
    Q = X * attn.Wq
    K = X * attn.Wk
    V = X * attn.Wv
    return scaled_dot_product_attention(Q, K, V)
end


struct MultiHeadSelfAttention
    heads::Vector{SelfAttention}
    Wo::Matrix{Float64}
    n_heads::Int
end


function MultiHeadSelfAttention(d_model::Int, n_heads::Int; seed::Int=42)
    @assert n_heads > 0 "n_heads must be > 0"
    @assert d_model > 0 "d_model must be > 0"
    @assert d_model % n_heads == 0 "d_model must be divisible by n_heads"
    dk = d_model ÷ n_heads
    dv = d_model ÷ n_heads
    heads = [SelfAttention(d_model, dk, dv; seed=seed + i) for i in 1:n_heads]
    rng = MersenneTwister(seed + n_heads + 1)
    scale = sqrt(2.0 / (d_model + d_model))
    Wo = scale .* randn(rng, n_heads * dv, d_model)
    return MultiHeadSelfAttention(heads, Wo, n_heads)
end


function forward(mha::MultiHeadSelfAttention, X::Matrix{Float64})
    head_outputs = Matrix{Float64}[]
    weights_per_head = Matrix{Float64}[]
    for head in mha.heads
        out, w = forward(head, X)
        push!(head_outputs, out)
        push!(weights_per_head, w)
    end
    concat = hcat(head_outputs...)
    return concat * mha.Wo, weights_per_head
end


function print_attention_matrix(weights::Matrix{Float64}, tokens::Vector{String})
    print("\n      ")
    for token in tokens
        @printf("%6s", token)
    end
    println()
    for i in 1:length(tokens)
        @printf("%6s", tokens[i])
        for j in 1:length(tokens)
            @printf("%6.3f", weights[i, j])
        end
        println()
    end
end


function ascii_heatmap(weights::Matrix{Float64}, tokens::Vector{String};
                       chars::String=" .:-=+*#%@")
    print("\n      ")
    for t in tokens
        @printf("%6s", t)
    end
    println()
    w_max = maximum(weights)
    for i in 1:length(tokens)
        @printf("%6s", tokens[i])
        for j in 1:length(tokens)
            level = Int(floor(weights[i, j] * (length(chars) - 1) / w_max))
            level = min(level, length(chars) - 1)
            ch = chars[level + 1]
            @printf("    %s ", ch)
        end
        println()
    end
end


function demo_softmax_stability()
    println("\n" * "=" ^ 60)
    println("SOFTMAX NUMERIC STABILITY")
    println("=" ^ 60)
    logits = reshape([2.0, 1.0, 0.1], 1, 3)
    probs = softmax_rows(logits)
    @printf("\nLogits:  [%s]\n", join([@sprintf("%.4f", v) for v in logits], ", "))
    @printf("Softmax: [%s]\n", join([@sprintf("%.4f", v) for v in probs], ", "))
    @printf("Sum:     %.4f\n", sum(probs))

    big_logits = reshape([100.0, 200.0, 300.0], 1, 3)
    big_probs = softmax_rows(big_logits)
    @printf("\nLarge logits:  [%s]\n",
            join([@sprintf("%.1f", v) for v in big_logits], ", "))
    @printf("Softmax:       [%s]\n",
            join([@sprintf("%.4f", v) for v in big_probs], ", "))
    @printf("Sum:           %.4f\n", sum(big_probs))
    println("(no overflow because we subtract the row maximum before exp)")
end


function demo_self_attention()
    println("=" ^ 60)
    println("SELF-ATTENTION FROM SCRATCH")
    println("=" ^ 60)

    tokens = ["The", "cat", "sat", "on", "the", "mat"]
    n_tokens = length(tokens)
    d_model = 16
    dk = 8
    dv = 8

    rng = MersenneTwister(42)
    X = randn(rng, n_tokens, d_model)

    @printf("\nSentence: %s\n", join(tokens, " "))
    @printf("Tokens: %d  d_model: %d  dk: %d  dv: %d\n", n_tokens, d_model, dk, dv)
    @printf("Input shape: (%d, %d)\n", size(X, 1), size(X, 2))

    attn = SelfAttention(d_model, dk, dv; seed=42)
    output, weights = forward(attn, X)
    @printf("\nOutput shape: (%d, %d)\n", size(output, 1), size(output, 2))
    println("\nAttention weights:")
    print_attention_matrix(weights, tokens)
    println("\nASCII heatmap (denser char = higher attention):")
    ascii_heatmap(weights, tokens)
    return tokens, X, d_model
end


function demo_multi_head(tokens::Vector{String}, X::Matrix{Float64}, d_model::Int)
    println("\n" * "=" ^ 60)
    println("MULTI-HEAD SELF-ATTENTION")
    println("=" ^ 60)
    n_heads = 2
    mha = MultiHeadSelfAttention(d_model, n_heads; seed=42)
    out, head_weights = forward(mha, X)
    @printf("\nHeads: %d  Output shape: (%d, %d)\n",
            n_heads, size(out, 1), size(out, 2))
    for (h, w) in enumerate(head_weights)
        @printf("\nHead %d attention weights:\n", h)
        print_attention_matrix(w, tokens)
    end
end


function main()
    tokens, X, d_model = demo_self_attention()
    demo_multi_head(tokens, X, d_model)
    demo_softmax_stability()
end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: phases/07-transformers-deep-dive/02-self-attention-from-scratch/code/main.rs
// Self-attention kernel from scratch, stdlib only.
// Topic: scaled dot-product attention with explicit row-major memory.
// References (cited in spirit, not as deps):
//   - Vaswani 2017, "Attention Is All You Need": https://arxiv.org/abs/1706.03762
//   - candle reference attention kernel:        https://github.com/huggingface/candle/blob/main/candle-nn/src/ops.rs
//   - Karpathy llm.c attention forward pass:    https://github.com/karpathy/llm.c/blob/master/train_gpt2.c
//
// Compile + run:  rustc --edition 2021 main.rs -o /tmp/sa && /tmp/sa

use std::f32::consts::E;

// Row-major matrix backed by a flat Vec<f32>. Helpers index by (row, col).
struct Mat {
    rows: usize,
    cols: usize,
    data: Vec<f32>,
}

impl Mat {
    fn zeros(rows: usize, cols: usize) -> Self {
        Mat { rows, cols, data: vec![0.0; rows * cols] }
    }

    #[inline] fn at(&self, i: usize, j: usize) -> f32 { self.data[i * self.cols + j] }
    #[inline] fn set(&mut self, i: usize, j: usize, v: f32) { self.data[i * self.cols + j] = v; }

    fn matmul(&self, b: &Mat) -> Mat {
        assert_eq!(self.cols, b.rows, "shape mismatch: {}x{} @ {}x{}", self.rows, self.cols, b.rows, b.cols);
        let mut out = Mat::zeros(self.rows, b.cols);
        for i in 0..self.rows {
            for k in 0..self.cols {
                let aik = self.at(i, k);
                if aik == 0.0 { continue; }
                let row_base = i * out.cols;
                let bk_base = k * b.cols;
                for j in 0..b.cols {
                    out.data[row_base + j] += aik * b.data[bk_base + j];
                }
            }
        }
        out
    }

    fn transpose(&self) -> Mat {
        let mut t = Mat::zeros(self.cols, self.rows);
        for i in 0..self.rows {
            for j in 0..self.cols {
                t.set(j, i, self.at(i, j));
            }
        }
        t
    }

    fn scale(&mut self, s: f32) {
        for v in self.data.iter_mut() { *v *= s; }
    }
}

// Softmax along the last axis (per row), numerically stable.
fn softmax_rows(m: &Mat) -> Mat {
    let mut out = Mat::zeros(m.rows, m.cols);
    for i in 0..m.rows {
        let mut row_max = f32::NEG_INFINITY;
        for j in 0..m.cols { if m.at(i, j) > row_max { row_max = m.at(i, j); } }
        let mut sum = 0.0f32;
        for j in 0..m.cols {
            let e = E.powf(m.at(i, j) - row_max);
            out.set(i, j, e);
            sum += e;
        }
        let inv = 1.0 / sum;
        for j in 0..m.cols {
            let v = out.at(i, j) * inv;
            out.set(i, j, v);
        }
    }
    out
}

// Q @ K^T / sqrt(d_k), softmax, then @ V.
fn scaled_dot_product_attention(q: &Mat, k: &Mat, v: &Mat) -> (Mat, Mat) {
    let dk = q.cols as f32;
    let k_t = k.transpose();
    let mut scores = q.matmul(&k_t);
    scores.scale(1.0 / dk.sqrt());
    let weights = softmax_rows(&scores);
    let out = weights.matmul(v);
    (out, weights)
}

// Deterministic, dependency-free Gaussian via Box-Muller from a Lehmer LCG.
struct Rng { state: u64 }
impl Rng {
    fn new(seed: u64) -> Self { Rng { state: seed.wrapping_mul(0x9E37_79B9_7F4A_7C15) | 1 } }
    fn next_u32(&mut self) -> u32 {
        self.state = self.state.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        (self.state >> 33) as u32
    }
    fn uniform(&mut self) -> f32 {
        (self.next_u32() as f32 + 1.0) / (u32::MAX as f32 + 2.0)
    }
    fn gauss(&mut self) -> f32 {
        let u1 = self.uniform();
        let u2 = self.uniform();
        (-2.0 * u1.ln()).sqrt() * (2.0 * std::f32::consts::PI * u2).cos()
    }
}

fn randn(rows: usize, cols: usize, scale: f32, rng: &mut Rng) -> Mat {
    let mut m = Mat::zeros(rows, cols);
    for v in m.data.iter_mut() { *v = rng.gauss() * scale; }
    m
}

struct SelfAttention {
    wq: Mat,
    wk: Mat,
    wv: Mat,
}

impl SelfAttention {
    fn new(d_model: usize, dk: usize, dv: usize, rng: &mut Rng) -> Self {
        let s_qk = (2.0 / (d_model + dk) as f32).sqrt();
        let s_v = (2.0 / (d_model + dv) as f32).sqrt();
        SelfAttention {
            wq: randn(d_model, dk, s_qk, rng),
            wk: randn(d_model, dk, s_qk, rng),
            wv: randn(d_model, dv, s_v, rng),
        }
    }

    fn forward(&self, x: &Mat) -> (Mat, Mat) {
        let q = x.matmul(&self.wq);
        let k = x.matmul(&self.wk);
        let v = x.matmul(&self.wv);
        scaled_dot_product_attention(&q, &k, &v)
    }
}

fn print_attention(weights: &Mat, tokens: &[&str]) {
    print!("      ");
    for t in tokens { print!("{:>7}", t); }
    println!();
    for i in 0..weights.rows {
        print!("{:>6}", tokens[i]);
        for j in 0..weights.cols { print!("{:>7.3}", weights.at(i, j)); }
        println!();
    }
}

fn ascii_heatmap(weights: &Mat, tokens: &[&str]) {
    let chars = [' ', '\u{2591}', '\u{2592}', '\u{2593}', '\u{2588}'];
    let mut w_max = 0.0f32;
    for v in &weights.data { if *v > w_max { w_max = *v; } }
    print!("      ");
    for t in tokens { print!("{:>7}", t); }
    println!();
    for i in 0..weights.rows {
        print!("{:>6}", tokens[i]);
        for j in 0..weights.cols {
            let level = ((weights.at(i, j) * (chars.len() - 1) as f32) / w_max) as usize;
            let level = level.min(chars.len() - 1);
            print!("     {} ", chars[level]);
        }
        println!();
    }
}

fn softmax_vec(logits: &[f32]) -> Vec<f32> {
    let mut m = f32::NEG_INFINITY;
    for &x in logits { if x > m { m = x; } }
    let exps: Vec<f32> = logits.iter().map(|x| (x - m).exp()).collect();
    let s: f32 = exps.iter().sum();
    exps.into_iter().map(|x| x / s).collect()
}

fn main() {
    let sentence = ["The", "cat", "sat", "on", "the", "mat"];
    let n_tokens = sentence.len();
    let d_model: usize = 16;
    let dk: usize = 8;
    let dv: usize = 8;

    println!("{}", "=".repeat(60));
    println!("SELF-ATTENTION FROM SCRATCH (Rust port)");
    println!("{}", "=".repeat(60));

    let mut rng = Rng::new(42);
    let x = randn(n_tokens, d_model, 1.0, &mut rng);
    println!("\nSentence: {}", sentence.join(" "));
    println!("Tokens: {}, d_model: {}, dk: {}, dv: {}", n_tokens, d_model, dk, dv);
    println!("Input shape: ({}, {})", x.rows, x.cols);

    let mut rng_w = Rng::new(42);
    let attn = SelfAttention::new(d_model, dk, dv, &mut rng_w);
    let (out, weights) = attn.forward(&x);

    println!("\nOutput shape: ({}, {})", out.rows, out.cols);
    println!("\nAttention weights:");
    print_attention(&weights, &sentence);

    println!("\nASCII heatmap (darker = higher attention):");
    ascii_heatmap(&weights, &sentence);

    println!("\n{}", "=".repeat(60));
    println!("SOFTMAX DEMO");
    println!("{}", "=".repeat(60));

    let logits = [2.0f32, 1.0, 0.1];
    let probs = softmax_vec(&logits);
    println!("\nLogits:  {:?}", logits);
    println!("Softmax: {:?}", probs.iter().map(|p| (p * 10000.0).round() / 10000.0).collect::<Vec<_>>());
    println!("Sum:     {:.4}", probs.iter().sum::<f32>());

    let large = [100.0f32, 200.0, 300.0];
    let probs_l = softmax_vec(&large);
    println!("\nLarge logits:  {:?}", large);
    println!("Softmax:       {:?}", probs_l.iter().map(|p| (p * 10000.0).round() / 10000.0).collect::<Vec<_>>());
    println!("Sum:           {:.4}", probs_l.iter().sum::<f32>());
    println!("(numerically stable, no overflow)");

    println!("\n{}", "=".repeat(60));
    println!("MICROBENCH: 10K attention forwards");
    println!("{}", "=".repeat(60));
    let start = std::time::Instant::now();
    let mut sink = 0.0f32;
    for _ in 0..10_000 {
        let (o, _) = attn.forward(&x);
        sink += o.at(0, 0);
    }
    let elapsed = start.elapsed();
    println!("10K forwards in {:.2}ms ({:.0}/sec)  sink={:.4}",
        elapsed.as_secs_f64() * 1000.0,
        10_000.0 / elapsed.as_secs_f64(),
        sink,
    );
}
